File: flood-fill (dfs)/flood-fill.go
package main

import "fmt"

type Cell struct {
	Row    int
	Column int
}

func floodFill(grid [][]int, hasVisited [][]bool) {
	row := len(grid)
	column := len(grid[0])

	stack := []Cell{}
	stack = append(stack, Cell{Row: 0, Column: 0})

	for len(stack) > 0 {
		lastIdx := len(stack) - 1
		current := stack[lastIdx]
		stack = stack[:lastIdx]

		x := current.Row
		y := current.Column
		if x >= 0 && x < row && y >= 0 && y < column && !hasVisited[x][y] {
			hasVisited[x][y] = true
			stack = append(stack, Cell{x - 1, y})
			stack = append(stack, Cell{x, y + 1})
			stack = append(stack, Cell{x + 1, y})
			stack = append(stack, Cell{x, y - 1})
		}
	}
}

func main() {
	row := 10
	column := 20
	value := 1

	grid := make([][]int, row)
	hasVisited := make([][]bool, row)
	for i := range grid {
		grid[i] = make([]int, column)
		hasVisited[i] = make([]bool, column)
		for j := range grid[i] {
			grid[i][j] = value
			value++
		}
	}

	floodFill(grid, hasVisited)

	fmt.Println(hasVisited)
}

File: flood-fill (dfs)/flood-fill.rb
def flood_fill(grid, has_visited)
    row = grid.size
    column = grid[0].size
    
    stack = []
    stack << [0, 0]

    while stack.size > 0 do
        current = stack.pop
        
        x = current[0]
        y = current[1]
        if x >= 0 && x < row && y >= 0 && y < column && !has_visited[x][y]
            has_visited[x][y] = true

            stack << [x-1, y]
            stack << [x, y+1]
            stack << [x+1, y]
            stack << [x, y-1]
        end
    end

    puts has_visited.inspect
end

def create_grid(row, column)
    grid = []
    (0...row).each do |i|
        grid << []
        (0...column).each do |j|
            grid[i] << 10*i + j
        end
    end
    grid
end

def create_flag(row, column)
    Array.new(row){Array.new(column){false}}
end

grid = create_grid(10, 20)
has_visited = create_flag(10, 20)
flood_fill(grid, has_visited)